engines/syguspdr.cpp

```diff
@@ -0,0 +1,1542 @@
+    assert(conjvec_partial.empty());
+    conjvec_partial.push_back(solver_true_);
+  }
+  syntax_analysis::IC3FormulaModel * partial_model = 
+    new syntax_analysis::IC3FormulaModel(std::move(cube_partial), conj_partial);
+
+  assert(partial_model);
+  model2cube_.emplace(conj_partial, partial_model);
+
+  return std::make_pair(
+    IC3Formula(conj_partial, conjvec_partial, false /*not a disjunction*/ ),
+    partial_model
+  );
+} // SygusPdr::ExtractPartialAndFullModel
+
+
+void SygusPdr::predecessor_generalization(size_t i, const Term & cterm, IC3Formula & pred) {
+  // used in rel_ind_check
+  // extract the model based on var list
+  // NOTE: i may be incorrect, it is given as F/\T->(here is i)
+
+  // no need to pop (pop in rel_ind_check)
+  // return the model and build IC3FormulaModel
+  auto partial_full_model = ExtractPartialModel(cterm);
+  pred = partial_full_model.first;
+} // generalize_predecessor
+
+
+bool SygusPdr::keep_var_in_partial_model(const Term & v) const {
+  if (options_.sygus_use_operator_abstraction_) {
+    assert (op_abstractor_ != nullptr);
+    const auto & dummy_inputs = op_abstractor_->dummy_inputs();
+    if (dummy_inputs.find(v) != dummy_inputs.end())
+      return true;
+  }
+
+  if (has_assumptions) { // must keep input vars
+    if(custom_ts_->is_curr_var(v) || custom_ts_->inputvars().find(v) != custom_ts_->inputvars().end() )
+      return true;
+    return false;
+  }
+  return custom_ts_->is_curr_var(v);
+} // keep_var_in_partial_model
+
+
+
+void SygusPdr::abstract() {
+  // called in initialize()
+} // SygusPdr::abstract()
+
+
+RefineResult SygusPdr::refine() {
+  if (!options_.sygus_use_operator_abstraction_)
+    return REFINE_NONE;
+
+  assert(witness_length()>0);
+
+  TermVec new_constraints;
+  bool succ = op_abstractor_->refine_with_constraints(
+    cex_, bad_, new_constraints);
+  assert(succ == (!new_constraints.empty()));
+  if (succ) {
+    for (const auto & c : new_constraints) {
+      ts_.add_constraint(c, true);
+      custom_ts_->add_constraint(c, true);
+      D(4, "[Refine] : {} ", c->to_string());
+
+      // rework the constraints buffers
+      if (custom_ts_->no_next(c)) {
+        constraints_curr_var_.push_back(c);
+        // translate input_var to next input_var
+        // but the state var ...
+        constraints_curr_var_.push_back(
+          custom_ts_->to_next_func(
+            solver_->substitute(c,
+              custom_ts_->input_var_to_next_map())));
+      } // else skip
+    }
+    reset_solver(); // because tr/init could be different now
+
+    // rework the constraints
+    has_assumptions = ! (custom_ts_->constraints().empty());
+
+    return REFINE_SUCCESS;
+  } // else
+  return REFINE_NONE;
+} // SygusPdr::refine()
+
+
+// ------------------------------------------------------------------
+//  May Block 
+
+/**
+ * Priority queue of proof obligations inspired by open-source ic3ia
+ * implementation
+ */
+class MayProofGoalQueue
```

Comment:
This is the same, right? If you need access to it here, we can just expose it in the base class. But I'd prefer to just make changes to the base class so that you don't need access.